src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs

```diff
@@ -8,5 +14,58 @@ protected override BoundExpression VisitExpressionWithoutStackGuard(BoundExpress
         {
             return (BoundExpression)Visit(node);
         }
+
+        public override BoundNode VisitBinaryOperator(BoundBinaryOperator node)
+        {
+            return VisitBinaryOperatorBase(node);
+        }
+
+        public override BoundNode VisitUserDefinedConditionalLogicalOperator(BoundUserDefinedConditionalLogicalOperator node)
+        {
+            return VisitBinaryOperatorBase(node);
+        }
+
+        private BoundNode VisitBinaryOperatorBase(BoundBinaryOperatorBase binaryOperator)
+        {
+            // Use an explicit stack to avoid blowing the managed stack when visiting deeply-recursive
+            // binary nodes
+            var stack = ArrayBuilder<BoundBinaryOperatorBase>.GetInstance();
+            BoundBinaryOperatorBase currentBinary = binaryOperator;
+
+            do
+            {
+                stack.Push(currentBinary);
+                currentBinary = currentBinary.Left as BoundBinaryOperatorBase;
+            } while (currentBinary != null);
+
+            Debug.Assert(stack.Count > 0);
+            var leftChild = (BoundExpression)Visit(stack.Peek().Left);
+
+            do
+            {
+                currentBinary = stack.Pop();
+
+                bool foundInfo = _updatedNullabilities.TryGetValue(currentBinary, out (NullabilityInfo Info, TypeSymbol Type) infoAndType);
+                var right = (BoundExpression)Visit(currentBinary.Right);
+                var type = foundInfo ? infoAndType.Type : currentBinary.Type;
+
+                currentBinary = currentBinary switch
+                {
+                    BoundBinaryOperator binary => (BoundBinaryOperatorBase)binary.Update(binary.OperatorKind, binary.ConstantValueOpt, binary.MethodOpt, binary.ResultKind, leftChild, right, type),
```

Comment:
>binary.MethodOpt [](start = 135, length = 16)

It feels like we might want to adjust the symbol as well.

Reply:
We definitely will, but we're not currently doing that anywhere. We'll need to do this for pretty much any bound node that has an internal symbol, like BoundCalls, BoundPropertyAccesses, etc.

---
In reply to: [271540045](https://github.com/dotnet/roslyn/pull/34429#discussion_r271540045) [](ancestors = 271540045)

Reply:
> We definitely will, but we're not currently doing that anywhere. We'll need to do this for pretty much any bound node that has an internal symbol, like BoundCalls, BoundPropertyAccesses, etc.

Consider adding a prototype comment

---
In reply to: [273229265](https://github.com/dotnet/roslyn/pull/34429#discussion_r273229265) [](ancestors = 273229265,271540045)